server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -745,9 +746,8 @@ public void deleteStaleSegments(int lastNMetadataFilesToKeep) throws IOException
             return;
         }
 
-        List<String> metadataFilesEligibleToDelete = sortedMetadataFileList.subList(
-            lastNMetadataFilesToKeep,
-            sortedMetadataFileList.size()
+        List<String> metadataFilesEligibleToDelete = new ArrayList<>(
```

Comment:
How about moving the `sortedMetadataFileList.removeAll()` call to after the debug log statement to avoid making another copy of the sublist?

Reply:
The `sublist(...)` result seems like the source of the misbehavior, I'd rather isolate it and prevent potential future misuse instead of changing the order of operations to make the error less likely.  I've added a test case that exercises the logger and validates the output to confirm this works as I'd expect.